File: java/543_Diameter_of_Binary_Tree.java
/*
https://leetcode.com/problems/diameter-of-binary-tree/
Given a binary tree, you need to compute the length of the diameter of the tree. The diameter of a binary tree is the
length of the longest path between any two nodes in a tree. This path may or may not pass through the root.

For each node, update max diameter as sum of depths of left and right subtrees.
Time - O(n)
Space - O(1)
*/

class Solution {
    int diameter = 0;

    public int diameterOfBinaryTree(TreeNode root) {
        depth(root);
        return diameter;
    }

    private int depth(TreeNode node) {  // returns depth of deepest leaf, leaf node has depth of 1
        if (node == null)
            return 0;

        int left = depth(node.left);
        int right = depth(node.right);

        diameter = Math.max(diameter, left + right);
        return Math.max(left + 1, right + 1);
    }
}

File: python_1_to_1000/543_Diameter_of_Binary_Tree.py
_author_ = 'jake'
_project_ = 'leetcode'

# https://leetcode.com/problems/diameter-of-binary-tree/
# Given a binary tree, you need to compute the length of the diameter of the tree.
# The diameter of a binary tree is the length of the longest path between any two nodes in a tree.
# This path may or may not pass through the root.

# Bottom up recursion.
# The longest path passing with a node as root is the longest downwards left path + longest downwards right path.
# If there is no child on right or left then the longest path on that side is zero.
# Else longest path is 1 + longest of left and right paths down from child.
# Time - O(n)
# Space - O(1)

class Solution(object):
    def diameterOfBinaryTree(self, root):
        """
        :type root: TreeNode
        :rtype: int
        """
        self.result = 0

        def helper(node):       # return longest downwards path
            if not node:
                return -1       # so leaf returns zero
            left = helper(node.left)
            right = helper(node.right)
            self.result = max(self.result, 2 + left + right)
            return max(1 + left, 1 + right)

        helper(root)        # ignore return value
        return self.result